### DISCUSSION TITLE
Scaling output vectors?

### DISCUSSION THREAD

**Discussion by EvanKomp** [e3nn/e3nn#491] [Q&A]
Hey all.

With a typical scalar output, a lot of the time we train the NN to predict the unit norm scaled labels, such that the neuron weights need not get too far from 0.0.

How should I approach this for vector outputs? I am training a model to reconstruct positions of atoms based on noise. I am applying noise if width half the average bond length in my large dataset, and asking the model to predict the inverse of that vector. This is oversimplified, but something like this:

```
positions # cartesian points
noise = torch.randn_like(positions) * BOND_LENGTH
positions_noised = positions + noise
noise_targets = -noise

outputs = model(....) # output Irreps 1x1o

loss = torch.mean(torch.square(torch.norm(outputs - noise_targets)))
```

Should I instead noise at bond length but compute loss at norm 1? eg.

```
positions # cartesian points
noise = torch.randn_like(positions)
positions_noised = positions + (noise * BOND_LENGTH)
noise_targets = -noise

outputs = model(....) # output Irreps 1x1o

loss = torch.mean(torch.square(torch.norm(outputs - noise_targets)))
```

Thanks for any wisdom.

**Comment by ameya98** [comment]
Hi!
I actually discussed exactly this in one of my recent papers :)
Please see Appendix B of https://arxiv.org/html/2410.14621v1#S3 where we train a equivariant denoiser model for point clouds. 
The idea is to normalize the effective inputs, outputs and loss at all noise levels.
The network interpolates between noise prediction (at low noise levels) and true data prediction (at high noise levels).
Happy to clarify anything, and hope this helps!

**Reply by EvanKomp** [reply]
This is great context, thank you.

If I am understanding correctly your proposal is to normalize as a function of $C \propto ||x_i - x_j||$ as opposed to normalizing directly to the noise vectors applied?

The application in practice and its benefit are clear to me re. input and output - but I am a bit fuzzier on scaling within the network and the skip connection. Am I wrong in my interpretation that this primarily enables the model to adapt to noise levels without requiring independant weights for each noise level?

With this in mind, for my application where I am intending to denoise a single step (along with atom masking) as a pretraining strategy  but **do not** intend to create an actual denoising network, I am thinking that input and output normalization is sufficient? And if so: does not including a skip connection and internal noise dependent normalization change the derivation of cin and cout?

Thanks for your time and hopefully this question finds you well.

**Reply by ameya98** [reply]
Hmm, can you be a bit more precise in terms of your task?
The 'explicit' skip gets scaled by $c_\text{skip}$ but internally the module $F_\theta$ can also have skip connections.

> Am I wrong in my interpretation that this primarily enables the model to adapt to noise levels without requiring independant weights for each noise level?

Yes, the idea is to remove the noise-dependent scaling factors. The network still needs some information about $\sigma$ for the denoising (otherwise, the denoising task is not well-defined in the first place).

**Reply by EvanKomp** [reply]
Thanks for the reply.

I am creating a modeler of portions of biopolymer systems. Eg. not a full protein. The purpose of the model is not to run dynamics or generate structures, but instead 1) embed these substructures to be used for clustering or 2) downstream predictions/finetuning. I suppose down the line it might be useful to use the denoising to perturb atoms in ur systems but not a primary goal.

The purpose of the denoising task at this point is strictly to help train a foundational model with understanding of geometric constraints. The components I wrote indeed have some internal skip connections over attention layers but I had not intended to skip the input noised structure to the output. Further, since the goal of the model is not actually a denoiser - I was intending to only noise some fraction of atoms at a time and compute loss on those.

Given that I am only doing this once - how stupid would it be to assume that the network can learn $\sigma$ parametrically since it will not be a variable in the task? Eg. noise at some sigma and give that to the model. Ask the model to predict a vector with standard norm magnitude, multiply the output by the original $\sigma$ to compute loss.

**Reply by ameya98** [reply]
In theory, the network can learn the scaling factor, yes. We have found that normalizing appropriately helps the convergence, but of course, it is not necessary.

**Reply by ameya98** [reply]
Also, if you are using the Adam optimizer (for example), scaling the loss by a constant does not change the training dynamics in any way (because the estimates of $E[\nabla_\theta l]$ and $\sqrt{E[\nabla_\theta l^2]}$ also get scaled by this constant).
